content/developer/tutorials/server_framework_101/05_connect_the_dots.rst

```diff
@@ -576,23 +575,130 @@ the :code:`@api.onchange()` decorator. These methods are triggered when the spec
 are altered. They operate on the in-memory representation of a single-record recordset received
 through `self`. If field values are modified, the changes are automatically reflected in the UI.
 
-.. todo: explain the env (self.env.uid, self.env.user, self.env.ref(xml_id), self.env[model_name])
+.. example::
+   In the following example, onchange methods are implemented to:
+
+   - unpublish products when all sellers are removed;
+   - warn the user if changing the sales price would result in a negative margin;
+   - raise a blocking user error if the category is changed after sales have been made.
+
+   .. code-block:: python
+
+      from odoo import _, api, fields, models
+      from odoo.exceptions import UserError
+
+
+      class Product(models.Model):
+          is_published = fields.Boolean(string="Published")
+
+          @api.onchange('seller_ids')
+          def _onchange_seller_ids_unpublish_if_no_sellers(self):
+              if not self.seller_ids:
+                  self.is_published = False
+
+          @api.onchange('price')
+          def _onchange_price_warn_if_margin_is_negative(self):
+              if self.margin < 0:
+                  return {
+                      'warning': {
+                          'title': _("Warning"),
+                          'message': _(
+                              "The sales price was changed from %(before_price)s to %(new_price)s, which"
+                              " would result in a negative margin. A sales price of minimum %(min_price)s"
+                              " is recommended.",
+                              before_price=self._origin.price, new_price=self.price, min_price=self.cost,
+                          ),
+                      }
+                  }
+
+          @api.onchange('category_id')
+          def _onchange_category_id_block_if_existing_sales(self):
+              existing_sales = self.env['sales.order'].search([('product_id', '=', self._origin.id)])
+              if existing_sales:
+                  raise UserError(_(
+                      "You cannot change the category of a product that has already been sold; unpublish"
+                      " it instead."
+                  ))
+
+   .. note::
+      - It is recommended to give self-explanatory names to onchange methods as multiple onchange
+        methods can be defined for a single field.
+      - Onchange methods don't need to iterate over the records as `self` is always a recordset of
+        length 1.
+      - The :code:`_()` function from the `odoo` package marks display strings :dfn:`strings shown
```

Comment:
why not _lt ? :)

Reply:
That will be for chapter 30, not 5 :stuck_out_tongue_winking_eye: